frontend/apps/reader/modules/readerlink.lua

```diff
@@ -576,56 +628,60 @@ function ReaderLink:onGotoLink(link, neglect_current_location, allow_footnote_po
             })
         end
         return true
-    end
+    elseif self.ui.wallabag then
+        local settings = G_reader_settings:readSetting("external_link_action")
+        local choose_action
+        if settings == "pop-up" or settings == nil then
+            local buttons = {
+                {
+                    {
+                        text = _("Cancel"),
+                        callback = function()
+                            UIManager:close(choose_action)
+                        end,
+                    },
+                    {
+                        text = _("Add to Wallabag"),
+                        callback = function()
+                            self.ui.wallabag:addArticle(link_url)
+                            UIManager:close(choose_action)
+                        end,
+                    },
 
-    -- Check if it is a link to a local file
-    local linked_filename = link_url:gsub("^file:", "") -- remove local file protocol if any
-    local anchor
-    if linked_filename:find("?") then -- remove any query string (including any following anchor)
-        linked_filename, anchor = linked_filename:match("^(.-)(%?.*)$")
-    elseif linked_filename:find("#") then -- remove any anchor
-        linked_filename, anchor = linked_filename:match("^(.-)(#.*)$")
-    end
-    linked_filename  = ffiutil.joinPath(self.document_dir, linked_filename) -- get full path
-    linked_filename = ffiutil.realpath(linked_filename) -- clean full path from ./ or ../
-    if linked_filename and lfs.attributes(linked_filename, "mode") == "file" then
-        local DocumentRegistry = require("document/documentregistry")
-        local provider = DocumentRegistry:getProvider(linked_filename)
-        if provider then
-            -- Display filename with anchor or query string, so the user gets
-            -- this information and can manually go to the appropriate place
-            local display_filename = linked_filename
-            if anchor then
-                display_filename = display_filename .. anchor
+                },
+            }
+            if Device:openLink() == nil then
```

Comment:
Calling `Device:openLink()` to know if it's supported seems a bit twisted. It returns `false `when set to `no`, and returns `nil` when it is supported, but provided with some invalid input...
If there's no cleaner way, at least `if Device:openLink() ~= false then` ?

Reply:
Not defining it as `no` and testing `Device.openLink`?

Reply:
Dunno. Is there already functions not defined in generic/device.lua that are defined in others?
I thought the idea of them being `no` was some kind of API listing everything that can be available.
Or add Device:canOpenLink = yes|no ?

Reply:
I think it's mainly that you can safely call it and be sure it won't crash. API listing is more like a happy side effect.

Having a separate canOpenLink seems a touch inelegant but I suppose that'll have to be the way forward. :-)